### PR TITLE
Add parse support for multiple requirements after `where` separated by `and`

### PR DESCRIPTION
**Pull request by josh11b**
Follow on to #4275 that added `where` parse support.

### CODE REVIEW

File: toolchain/parse/handle_requirement.cpp [hunk 1]
```diff
@@ -10,6 +10,7 @@ namespace Carbon::Parse {
 
 auto HandleRequirementBegin(Context& context) -> void {
   context.PopAndDiscardState();
+  // TODO: Peek aheadd for `.designator = ...`, and give it special handling.
```
Comment (zygoloid) [inline toolchain/parse/handle_requirement.cpp]:
```suggestion
  // TODO: Peek ahead for `.designator = ...`, and give it special handling.
```

File: toolchain/parse/state.def [hunk 2]
```diff
@@ -645,6 +645,14 @@ CARBON_PARSE_STATE(RequirementOperator)
...
 //                        ^
 //   (state done)
+// expr where expr impls expr and
+//                            ^~~
+// expr where expr = expr and
+//                        ^~~
+// expr where expr == expr and
+//                         ^~~
+//   1. Expr
+//   2. RequirementOperator
```
Comment (zygoloid) [inline toolchain/parse/state.def]:
```suggestion
//   1. RequirementBegin
```

File: toolchain/lex/tokenized_buffer.h [hunk 3]
```diff
@@ -219,6 +219,8 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
 
   auto size() const -> int { return token_infos_.size(); }
 
+  // This is an upper bound on the number of output parse nodes in the absence
+  // of errors.
   auto expected_parse_tree_size() const -> int {
```
Comment (zygoloid) [inline toolchain/lex/tokenized_buffer.h]: Might be clearer to rename this to `expected_max_parse_tree_size` or similar.

**Review by zygoloid — APPROVED** [review]
Looks good; feel free to merge with the suggested changes applied.

File: toolchain/lex/tokenized_buffer.h (hunk 3 above)
Comment (josh11b) [reply toolchain/lex/tokenized_buffer.h]: Done.

**Pull request merged**